File: server/encrypted-object/service.ts
import { decodeOrThrow } from '../../lib/codec/core';
import type { VaultContext } from '../../lib/domain/identity';
import {
  decodeEnvelopeCiphertext,
  ENVELOPE_CRYPTO_VERSION,
  type CryptoObjectRevision,
  type EnvelopeCiphertext,
  type EnvelopeObject,
  type VaultDekKeyring,
} from '../crypto/core';
import type { EnvelopeEncryptionService } from '../crypto/envelope-service';
import {
  opaqueObjectKeyDecoder,
  planDeleteAttempt,
  planEncryptedObjectWrite,
  planOrphanCollection,
  planPendingWriteResume,
  planStoredCiphertext,
  storedByteCountDecoder,
  type EncryptedObjectMetadata,
  type EncryptedWriteId,
  type PendingEncryptedWrite,
} from './core';
import type {
  EncryptedObjectMetadataRepository,
  OpaqueObjectKeyGeneratorPort,
  PrivateObjectStoragePort,
} from './ports';

export class EncryptedObjectIntegrityError extends Error {
  constructor() {
    super('Encrypted object integrity validation failed');
    this.name = 'EncryptedObjectIntegrityError';
  }
}

export class EncryptedObjectStorageConflictError extends Error {
  constructor() {
    super('Immutable encrypted object already has different content');
    this.name = 'EncryptedObjectStorageConflictError';
  }
}

export type EncryptedObjectWriteResult =
  | {
      readonly kind: 'stored' | 'replayed';
      readonly metadata: EncryptedObjectMetadata;
    }
  | {
      readonly kind: 'not-applied';
      readonly reason:
        | 'idempotency-key-reuse'
        | 'unexpected-existing-object'
        | 'missing-object'
        | 'stale-revision'
        | 'invalid-next-revision'
        | 'invalid-timeline'
        | 'ciphertext-limit'
        | 'cas-conflict';
    };

export type EncryptedObjectReadResult =
  | { readonly kind: 'found'; readonly plaintext: Uint8Array }
  | { readonly kind: 'not-found' };

export type EncryptedObjectService = {
  write(input: {
    readonly object: EnvelopeObject;
    readonly expectedRevision: CryptoObjectRevision | null;
    readonly nextRevision: CryptoObjectRevision;
    readonly writeId: EncryptedWriteId;
    readonly plaintext: Uint8Array;
    readonly keyring: VaultDekKeyring;
    readonly createdAt: number;
  }): Promise<EncryptedObjectWriteResult>;
  read(input: {
    readonly object: EnvelopeObject;
    readonly keyring: VaultDekKeyring;
  }): Promise<EncryptedObjectReadResult>;
  readRevision(input: {
    readonly object: EnvelopeObject;
    readonly objectRevision: CryptoObjectRevision;
    readonly keyring: VaultDekKeyring;
  }): Promise<EncryptedObjectReadResult>;
  collectOrphans(input: {
    readonly scanStartedAt: number;
    readonly gracePeriodMs: number;
  }): Promise<{ readonly enqueued: number }>;
  drainDeleteOutbox(input: {
    readonly now: number;
    readonly retryDelayMs: number;
    readonly limit: number;
  }): Promise<{ readonly completed: number; readonly retried: number }>;
};

export function createEncryptedObjectService(input: {
  readonly context: VaultContext;
  readonly metadata: EncryptedObjectMetadataRepository;
  readonly objects: PrivateObjectStoragePort;
  readonly objectKeys: OpaqueObjectKeyGeneratorPort;
  readonly encryption: EnvelopeEncryptionService;
  readonly maximumCiphertextBytes?: number;
}): EncryptedObjectService {
  const maximumCiphertextBytes =
    input.maximumCiphertextBytes === undefined
      ? undefined
      : decodeOrThrow(
          storedByteCountDecoder,
          input.maximumCiphertextBytes,
          'encrypted object ciphertext limit',
        );
  return {
    async write(command) {
      const plaintextBytes = decodeOrThrow(
        storedByteCountDecoder,
        command.plaintext.byteLength,
        'encrypted object plaintext size',
      );
      const request = {
        object: command.object,
        expectedRevision: command.expectedRevision,
        nextRevision: command.nextRevision,
        writeId: command.writeId,
        plaintextBytes,
        dekVersion: command.keyring.writeVersion,
        createdAt: command.createdAt,
      } as const;
      const existingWrite = await input.metadata.findByWriteId(command.writeId);
      if (existingWrite !== undefined) {
        const plan = planEncryptedObjectWrite({
          existingWrite,
          current: undefined,
          request,
        });
        if (plan.kind === 'replay') {
          if (
            maximumCiphertextBytes !== undefined &&
            plan.metadata.ciphertextBytes > maximumCiphertextBytes
          ) {
            return { kind: 'not-applied', reason: 'ciphertext-limit' };
          }
          return { kind: 'replayed', metadata: plan.metadata };
        }
        if (plan.kind === 'rejected') {
          return { kind: 'not-applied', reason: plan.reason };
        }
        throw new EncryptedObjectIntegrityError();
      }

      const existingIntent = await input.metadata.findIntent(command.writeId);
      let intent: PendingEncryptedWrite;
      if (existingIntent !== undefined) {
        const resume = planPendingWriteResume(existingIntent, request);
        if (resume.kind === 'rejected') {
          return { kind: 'not-applied', reason: resume.reason };
        }
        intent = resume.intent;
      } else {
        const current = await input.metadata.findCurrent(command.object);
        const plan = planEncryptedObjectWrite({
          existingWrite: undefined,
          current,
          request,
        });
        if (plan.kind === 'rejected') {
          return { kind: 'not-applied', reason: plan.reason };
        }
        if (plan.kind === 'replay') throw new EncryptedObjectIntegrityError();
        const objectKey = decodeOrThrow(
          opaqueObjectKeyDecoder,
          await input.objectKeys.createObjectKey(),
          'opaque encrypted object key source',
        );
        const proposedIntent: PendingEncryptedWrite = {
          object: command.object,
          expectedRevision: command.expectedRevision,
          objectRevision: command.nextRevision,
          writeId: command.writeId,
          objectKey,
          plaintextBytes,
          cryptoVersion: ENVELOPE_CRYPTO_VERSION,
          dekVersion: command.keyring.writeVersion,
          createdAt: command.createdAt,
        };
        const reservation = await input.metadata.reserveIntent(proposedIntent);
        if (reservation.kind === 'conflict') {
          return { kind: 'not-applied', reason: 'cas-conflict' };
        }
        const resume = planPendingWriteResume(reservation.intent, request);
        if (resume.kind === 'rejected') {
          return { kind: 'not-applied', reason: resume.reason };
        }
        intent = resume.intent;
      }

      let storedBytes = await input.objects.get(intent.objectKey);
      let authenticateExistingObject = storedBytes !== undefined;
      if (storedBytes === undefined) {
        const ciphertext = await input.encryption.encrypt({
          keyring: command.keyring,
          context: {
            vaultId: input.context.vaultId,
            object: intent.object,
            objectRevision: intent.objectRevision,
          },
          plaintext: command.plaintext,
        });
        storedBytes = encodeEncryptedObjectCiphertext(ciphertext);
        if (
          maximumCiphertextBytes !== undefined &&
          storedBytes.byteLength > maximumCiphertextBytes
        ) {
          await input.metadata.abandonIntent({
            intent,
            requestedAt: command.createdAt,
          });
          return { kind: 'not-applied', reason: 'ciphertext-limit' };
        }
        const put = await input.objects.putIfAbsent({
          objectKey: intent.objectKey,
          bytes: storedBytes,
          createdAt: intent.createdAt,
        });
        if (put.kind === 'conflict') {
          throw new EncryptedObjectStorageConflictError();
        }
        if (put.kind === 'already-present') {
          const existingBytes = await input.objects.get(intent.objectKey);
          if (existingBytes === undefined)
            throw new EncryptedObjectIntegrityError();
          storedBytes = existingBytes;
          authenticateExistingObject = true;
        }
      }

      if (
        maximumCiphertextBytes !== undefined &&
        storedBytes.byteLength > maximumCiphertextBytes
      ) {
        await input.metadata.abandonIntent({
          intent,
          requestedAt: command.createdAt,
        });
        return { kind: 'not-applied', reason: 'ciphertext-limit' };
      }

      const ciphertext = decodeEncryptedObjectCiphertext(storedBytes);
      const storedPlan = planStoredCiphertext({
        metadata: intent,
        ciphertext,
        actualCiphertextBytes: storedBytes.byteLength,
      });
      if (storedPlan.kind === 'rejected') {
        throw new EncryptedObjectIntegrityError();
      }
      if (authenticateExistingObject) {
        let authenticatedPlaintext: Uint8Array;
        try {
          authenticatedPlaintext = await input.encryption.decrypt({
            keyring: command.keyring,
            context: {
              vaultId: input.context.vaultId,
              object: intent.object,
              objectRevision: intent.objectRevision,
            },
            ciphertext,
          });
        } catch {
          throw new EncryptedObjectIntegrityError();
        }
        if (!bytesEqual(authenticatedPlaintext, command.plaintext)) {
          throw new EncryptedObjectIntegrityError();
        }
      }
      const commit = await input.metadata.commitIntent({
        intent,
        ciphertextBytes: storedBytes.byteLength,
      });
      if (commit.kind === 'not-applied') {
        await input.metadata.abandonIntent({
          intent,
          requestedAt: command.createdAt,
        });
        return { kind: 'not-applied', reason: 'cas-conflict' };
      }
      return { kind: 'stored', metadata: commit.metadata };
    },

    async read(command) {
      const metadata = await input.metadata.findCurrent(command.object);
      return readMetadata(input, command.keyring, metadata);
    },

    async readRevision(command) {
      const metadata = await input.metadata.findRevision(
        command.object,
        command.objectRevision,
      );
      return readMetadata(input, command.keyring, metadata);
    },

    async collectOrphans(command) {
      const [stored, protectedKeys] = await Promise.all([
        input.objects.list(),
        input.metadata.listProtectedObjectKeys(),
      ]);
      const orphanKeys = planOrphanCollection({
        stored,
        protectedKeys,
        scanStartedAt: command.scanStartedAt,
        gracePeriodMs: command.gracePeriodMs,
      });
      for (const objectKey of orphanKeys) {
        await input.metadata.enqueueDelete({
          objectKey,
          requestedAt: command.scanStartedAt,
        });
      }
      return { enqueued: orphanKeys.length };
    },

    async drainDeleteOutbox(command) {
      const entries = await input.metadata.listReadyDeletes({
        now: command.now,
        limit: command.limit,
      });
      let completed = 0;
      let retried = 0;
      for (const entry of entries) {
        let succeeded = true;
        try {
          await input.objects.delete(entry.objectKey);
        } catch {
          succeeded = false;
        }
        const plan = planDeleteAttempt({
          entry,
          succeeded,
          attemptedAt: command.now,
          retryDelayMs: command.retryDelayMs,
        });
        if (plan.kind === 'complete') {
          await input.metadata.completeDelete(entry);
          completed += 1;
        } else {
          await input.metadata.rescheduleDelete(plan.entry);
          retried += 1;
        }
      }
      return { completed, retried };
    },
  };
}

async function readMetadata(
  input: {
    readonly context: VaultContext;
    readonly objects: PrivateObjectStoragePort;
    readonly encryption: EnvelopeEncryptionService;
  },
  keyring: VaultDekKeyring,
  metadata: EncryptedObjectMetadata | undefined,
): Promise<EncryptedObjectReadResult> {
  if (metadata === undefined) return { kind: 'not-found' };
  const bytes = await input.objects.get(metadata.objectKey);
  if (bytes === undefined) throw new EncryptedObjectIntegrityError();
  const ciphertext = decodeEncryptedObjectCiphertext(bytes);
  const plan = planStoredCiphertext({
    metadata,
    ciphertext,
    actualCiphertextBytes: bytes.byteLength,
  });
  if (plan.kind === 'rejected') throw new EncryptedObjectIntegrityError();
  const plaintext = await input.encryption.decrypt({
    keyring,
    context: {
      vaultId: input.context.vaultId,
      object: metadata.object,
      objectRevision: metadata.objectRevision,
    },
    ciphertext,
  });
  if (plaintext.byteLength !== metadata.plaintextBytes) {
    throw new EncryptedObjectIntegrityError();
  }
  return { kind: 'found', plaintext };
}

export function encodeEncryptedObjectCiphertext(
  ciphertext: EnvelopeCiphertext,
): Uint8Array {
  return new TextEncoder().encode(JSON.stringify(ciphertext));
}

function bytesEqual(left: Uint8Array, right: Uint8Array): boolean {
  return (
    left.byteLength === right.byteLength &&
    left.every((value, index) => value === right[index])
  );
}

export function decodeEncryptedObjectCiphertext(bytes: Uint8Array) {
  try {
    const candidate: unknown = JSON.parse(new TextDecoder().decode(bytes));
    return decodeEnvelopeCiphertext(candidate);
  } catch {
    throw new EncryptedObjectIntegrityError();
  }
}

File: backend/internal/encryptedobject/service.go
package encryptedobject

import (
	"context"
	"crypto/subtle"
	"encoding/json"
	"errors"

	"github.com/fukamu/notes/backend/internal/cryptocontent"
	"github.com/fukamu/notes/backend/internal/identity"
)

var (
	ErrInvalidOperation = errors.New("invalid encrypted object operation")
	ErrIntegrity        = errors.New("encrypted object integrity validation failed")
	ErrStorageConflict  = errors.New("immutable encrypted object already has different content")
)

type WriteResultKind string

const (
	WriteStored     WriteResultKind = "stored"
	WriteReplayed   WriteResultKind = "replayed"
	WriteNotApplied WriteResultKind = "not-applied"
)

type WriteResult struct {
	Kind     WriteResultKind
	Metadata Metadata
	Reason   RejectionReason
}

type ReadResult struct {
	Found     bool
	Plaintext []byte
}

type WriteCommand struct {
	Object           ObjectRef
	ExpectedRevision *cryptocontent.ObjectRevision
	NextRevision     cryptocontent.ObjectRevision
	WriteID          WriteID
	Plaintext        []byte
	Keyring          cryptocontent.VaultDEKKeyring
	CreatedAtMilli   int64
}

type Service struct {
	vaultID                identity.VaultID
	metadata               MetadataRepository
	objects                ObjectStoragePort
	objectKeys             ObjectKeyGeneratorPort
	encryption             EncryptionPort
	maximumCiphertextBytes *int64
}

func NewService(
	vaultID identity.VaultID,
	metadata MetadataRepository,
	objects ObjectStoragePort,
	objectKeys ObjectKeyGeneratorPort,
	encryption EncryptionPort,
	maximumCiphertextBytes *int64,
) (*Service, error) {
	if _, err := identity.ParseVaultID(string(vaultID)); err != nil || metadata == nil || objects == nil ||
		objectKeys == nil || encryption == nil {
		return nil, ErrInvalidOperation
	}
	if maximumCiphertextBytes != nil && !validByteCount(*maximumCiphertextBytes) {
		return nil, ErrInvalidOperation
	}
	var copiedLimit *int64
	if maximumCiphertextBytes != nil {
		value := *maximumCiphertextBytes
		copiedLimit = &value
	}
	return &Service{
		vaultID: vaultID, metadata: metadata, objects: objects, objectKeys: objectKeys,
		encryption: encryption, maximumCiphertextBytes: copiedLimit,
	}, nil
}

func (service *Service) Write(ctx context.Context, command WriteCommand) (WriteResult, error) {
	request := WriteRequest{
		Object: command.Object, ExpectedRevision: copyRevision(command.ExpectedRevision),
		NextRevision: command.NextRevision, WriteID: command.WriteID,
		PlaintextBytes: int64(len(command.Plaintext)), DEKVersion: command.Keyring.WriteVersion,
		CreatedAtMilli: command.CreatedAtMilli,
	}
	if service == nil || ValidateWriteRequest(request) != nil || service.metadata == nil || service.objects == nil ||
		service.objectKeys == nil || service.encryption == nil {
		return WriteResult{}, ErrInvalidOperation
	}
	if _, err := command.Keyring.SelectForWrite(service.vaultID); err != nil {
		return WriteResult{}, ErrInvalidOperation
	}

	existingWrite, err := service.metadata.FindByWriteID(ctx, command.WriteID)
	if err != nil {
		return WriteResult{}, err
	}
	if existingWrite != nil {
		plan := PlanWrite(existingWrite, nil, request)
		if plan.Kind == WritePlanReplay && plan.Metadata != nil {
			if service.exceedsLimit(plan.Metadata.CiphertextBytes) {
				return WriteResult{Kind: WriteNotApplied, Reason: ReasonCiphertextLimit}, nil
			}
			return WriteResult{Kind: WriteReplayed, Metadata: *plan.Metadata}, nil
		}
		if plan.Kind == WritePlanRejected {
			return WriteResult{Kind: WriteNotApplied, Reason: plan.Reason}, nil
		}
		return WriteResult{}, ErrIntegrity
	}

	existingIntent, err := service.metadata.FindIntent(ctx, command.WriteID)
	if err != nil {
		return WriteResult{}, err
	}
	var intent PendingWrite
	if existingIntent != nil {
		if !PendingMatchesRequest(*existingIntent, request) {
			return WriteResult{Kind: WriteNotApplied, Reason: ReasonIdempotencyKeyReuse}, nil
		}
		intent = *existingIntent
	} else {
		current, findErr := service.metadata.FindCurrent(ctx, command.Object)
		if findErr != nil {
			return WriteResult{}, findErr
		}
		plan := PlanWrite(nil, current, request)
		if plan.Kind == WritePlanRejected {
			return WriteResult{Kind: WriteNotApplied, Reason: plan.Reason}, nil
		}
		if plan.Kind != WritePlanAccepted {
			return WriteResult{}, ErrIntegrity
		}
		rawObjectKey, createErr := service.objectKeys.CreateObjectKey(ctx)
		if createErr != nil {
			return WriteResult{}, createErr
		}
		objectKey, parseErr := ParseObjectKey(rawObjectKey)
		if parseErr != nil {
			return WriteResult{}, ErrInvalidOperation
		}
		proposed := PendingWrite{
			Object: command.Object, ExpectedRevision: copyRevision(command.ExpectedRevision),
			ObjectRevision: command.NextRevision, WriteID: command.WriteID, ObjectKey: objectKey,
			PlaintextBytes: request.PlaintextBytes, CryptoVersion: cryptocontent.EnvelopeCryptoVersion,
			DEKVersion: request.DEKVersion, CreatedAtMilli: command.CreatedAtMilli,
		}
		reservation, reserveErr := service.metadata.ReserveIntent(ctx, proposed)
		if reserveErr != nil {
			return WriteResult{}, reserveErr
		}
		if reservation.Kind == IntentConflict {
			return WriteResult{Kind: WriteNotApplied, Reason: ReasonCASConflict}, nil
		}
		if (reservation.Kind != IntentReserved && reservation.Kind != IntentExisting) ||
			!PendingMatchesRequest(reservation.Intent, request) {
			if reservation.Kind == IntentReserved || reservation.Kind == IntentExisting {
				return WriteResult{Kind: WriteNotApplied, Reason: ReasonIdempotencyKeyReuse}, nil
			}
			return WriteResult{}, ErrIntegrity
		}
		intent = reservation.Intent
	}
	currentAfterReservation, err := service.metadata.FindCurrent(ctx, command.Object)
	if err != nil {
		return WriteResult{}, err
	}
	reservationPlan := PlanWrite(nil, currentAfterReservation, request)
	if reservationPlan.Kind != WritePlanAccepted {
		if err := service.metadata.AbandonIntent(ctx, intent, command.CreatedAtMilli); err != nil {
			return WriteResult{}, err
		}
		return WriteResult{Kind: WriteNotApplied, Reason: ReasonCASConflict}, nil
	}

	storedBytes, found, err := service.objects.Get(ctx, intent.ObjectKey)
	if err != nil {
		return WriteResult{}, err
	}
	authenticateExisting := found
	if !found {
		plaintextForEncryption := append([]byte(nil), command.Plaintext...)
		ciphertext, encryptErr := service.encryption.Encrypt(
			ctx,
			command.Keyring,
			service.objectContext(intent.Object, intent.ObjectRevision),
			plaintextForEncryption,
		)
		clear(plaintextForEncryption)
		if encryptErr != nil {
			return WriteResult{}, encryptErr
		}
		storedBytes, err = encodeCiphertext(ciphertext)
		if err != nil {
			return WriteResult{}, ErrIntegrity
		}
		if service.exceedsLimit(int64(len(storedBytes))) {
			if err := service.metadata.AbandonIntent(ctx, intent, command.CreatedAtMilli); err != nil {
				return WriteResult{}, err
			}
			return WriteResult{Kind: WriteNotApplied, Reason: ReasonCiphertextLimit}, nil
		}
		put, putErr := service.objects.PutIfAbsent(ctx, intent.ObjectKey, storedBytes, intent.CreatedAtMilli)
		if putErr != nil {
			return WriteResult{}, putErr
		}
		switch put {
		case PutStored:
		case PutAlreadyPresent:
			storedBytes, found, err = service.objects.Get(ctx, intent.ObjectKey)
			if err != nil {
				return WriteResult{}, err
			}
			if !found {
				return WriteResult{}, ErrIntegrity
			}
			authenticateExisting = true
		case PutConflict:
			return WriteResult{}, ErrStorageConflict
		default:
			return WriteResult{}, ErrIntegrity
		}
	}
	if service.exceedsLimit(int64(len(storedBytes))) {
		if err := service.metadata.AbandonIntent(ctx, intent, command.CreatedAtMilli); err != nil {
			return WriteResult{}, err
		}
		return WriteResult{Kind: WriteNotApplied, Reason: ReasonCiphertextLimit}, nil
	}
	ciphertext, err := cryptocontent.DecodeEnvelopeCiphertext(storedBytes)
	if err != nil || !StoredCiphertextMatches(
		intent.CryptoVersion, intent.DEKVersion, nil, ciphertext, int64(len(storedBytes)),
	) {
		return WriteResult{}, ErrIntegrity
	}
	if authenticateExisting {
		authenticated, decryptErr := service.encryption.Decrypt(
			ctx, command.Keyring, service.objectContext(intent.Object, intent.ObjectRevision), ciphertext,
		)
		matches := decryptErr == nil && constantTimeEqual(authenticated, command.Plaintext)
		clear(authenticated)
		if !matches {
			return WriteResult{}, ErrIntegrity
		}
	}
	commit, err := service.metadata.CommitIntent(ctx, intent, int64(len(storedBytes)))
	if err != nil {
		return WriteResult{}, err
	}
	if commit.Kind == MetadataNotApplied {
		if err := service.metadata.AbandonIntent(ctx, intent, command.CreatedAtMilli); err != nil {
			return WriteResult{}, err
		}
		return WriteResult{Kind: WriteNotApplied, Reason: ReasonCASConflict}, nil
	}
	if commit.Kind != MetadataApplied || ValidateMetadata(commit.Metadata) != nil {
		return WriteResult{}, ErrIntegrity
	}
	return WriteResult{Kind: WriteStored, Metadata: commit.Metadata}, nil
}

func (service *Service) Read(
	ctx context.Context,
	object ObjectRef,
	keyring cryptocontent.VaultDEKKeyring,
) (ReadResult, error) {
	if service == nil || ValidateObjectRef(object) != nil {
		return ReadResult{}, ErrInvalidOperation
	}
	metadata, err := service.metadata.FindCurrent(ctx, object)
	if err != nil {
		return ReadResult{}, err
	}
	return service.readMetadata(ctx, keyring, metadata)
}

func (service *Service) ReadRevision(
	ctx context.Context,
	object ObjectRef,
	revision cryptocontent.ObjectRevision,
	keyring cryptocontent.VaultDEKKeyring,
) (ReadResult, error) {
	if service == nil || ValidateObjectRef(object) != nil || validateRevision(revision) != nil {
		return ReadResult{}, ErrInvalidOperation
	}
	metadata, err := service.metadata.FindRevision(ctx, object, revision)
	if err != nil {
		return ReadResult{}, err
	}
	return service.readMetadata(ctx, keyring, metadata)
}

func (service *Service) CollectOrphans(
	ctx context.Context,
	scanStartedAt, gracePeriodMilli int64,
) (int, error) {
	if service == nil || !validTimestamp(scanStartedAt) || gracePeriodMilli < 0 ||
		gracePeriodMilli > identity.MaximumSafeInteger {
		return 0, ErrInvalidOperation
	}
	stored, err := service.objects.List(ctx)
	if err != nil {
		return 0, err
	}
	protected, err := service.metadata.ListProtectedObjectKeys(ctx)
	if err != nil {
		return 0, err
	}
	enqueued := 0
	for _, objectKey := range PlanOrphanCollection(stored, protected, scanStartedAt, gracePeriodMilli) {
		applied, enqueueErr := service.metadata.EnqueueDelete(ctx, objectKey, scanStartedAt)
		if enqueueErr != nil {
			return 0, enqueueErr
		}
		if applied {
			enqueued++
		}
	}
	return enqueued, nil
}

func (service *Service) DrainDeleteOutbox(
	ctx context.Context,
	now, retryDelayMilli int64,
	limit int,
) (completed, retried int, err error) {
	if service == nil || !validTimestamp(now) || retryDelayMilli < 0 ||
		retryDelayMilli > identity.MaximumSafeInteger || now > identity.MaximumSafeInteger-retryDelayMilli ||
		limit < 1 || limit > 100 {
		return 0, 0, ErrInvalidOperation
	}
	entries, err := service.metadata.ListReadyDeletes(ctx, now, limit)
	if err != nil {
		return 0, 0, err
	}
	for _, entry := range entries {
		result, deleteErr := service.objects.Delete(ctx, entry.ObjectKey)
		succeeded := deleteErr == nil && (result == DeleteDeleted || result == DeleteNotFound)
		planned, complete := PlanDeleteAttempt(entry, succeeded, now, retryDelayMilli)
		if complete {
			if err := service.metadata.CompleteDelete(ctx, entry); err != nil {
				return completed, retried, err
			}
			completed++
			continue
		}
		if err := service.metadata.RescheduleDelete(ctx, planned); err != nil {
			return completed, retried, err
		}
		retried++
	}
	return completed, retried, nil
}

func (service *Service) readMetadata(
	ctx context.Context,
	keyring cryptocontent.VaultDEKKeyring,
	metadata *Metadata,
) (ReadResult, error) {
	if metadata == nil {
		return ReadResult{Found: false}, nil
	}
	if ValidateMetadata(*metadata) != nil || service.exceedsLimit(metadata.CiphertextBytes) {
		return ReadResult{}, ErrIntegrity
	}
	stored, found, err := service.objects.Get(ctx, metadata.ObjectKey)
	if err != nil {
		return ReadResult{}, err
	}
	if !found {
		return ReadResult{}, ErrIntegrity
	}
	ciphertext, err := cryptocontent.DecodeEnvelopeCiphertext(stored)
	if err != nil || !StoredCiphertextMatches(
		metadata.CryptoVersion, metadata.DEKVersion, &metadata.CiphertextBytes,
		ciphertext, int64(len(stored)),
	) {
		return ReadResult{}, ErrIntegrity
	}
	plaintext, err := service.encryption.Decrypt(
		ctx, keyring, service.objectContext(metadata.Object, metadata.ObjectRevision), ciphertext,
	)
	if err != nil || int64(len(plaintext)) != metadata.PlaintextBytes {
		return ReadResult{}, ErrIntegrity
	}
	return ReadResult{Found: true, Plaintext: plaintext}, nil
}

func (service *Service) objectContext(object ObjectRef, revision cryptocontent.ObjectRevision) cryptocontent.ObjectContext {
	return cryptocontent.ObjectContext{
		VaultID: service.vaultID, Kind: object.Kind, ObjectID: object.ObjectID, ObjectRevision: revision,
	}
}

func (service *Service) exceedsLimit(bytes int64) bool {
	return service.maximumCiphertextBytes != nil && bytes > *service.maximumCiphertextBytes
}

func encodeCiphertext(value cryptocontent.EnvelopeCiphertext) ([]byte, error) {
	if cryptocontent.ValidateEnvelopeCiphertext(value) != nil {
		return nil, ErrIntegrity
	}
	return json.Marshal(value)
}

func constantTimeEqual(left, right []byte) bool {
	return len(left) == len(right) && subtle.ConstantTimeCompare(left, right) == 1
}

func copyRevision(value *cryptocontent.ObjectRevision) *cryptocontent.ObjectRevision {
	if value == nil {
		return nil
	}
	copyOfValue := *value
	return &copyOfValue
}
